Allow the optional passing of the config to use

halostatue opened the pull request:
In production mode, this library will work perfectly because we rely on machine IAM keys. In development mode, because I have multiple IAM keys that may be used in our multi-tenant application, I need to be able to provide the configuration. This has no impact on most people, but will make this library usable by our developers.

halostatue committed:
Allow the optional passing of the config to use

rmoorman commented:
@halostatue just out of curiosity, you have less multitenancy in production than in development? I would think that the a production application would be more likely to require configurability for multiple keys. Or do you have one deployment/machine for each tenant in production? But then why is it problematic to add aws configuration for your specific checkout?

Anyway, I really would like to have this configurable at runtime for multiple sets of aws credentials too.

As for the proposed changes, wouldn't it be a bit cleaner to have a function with arity 1 for defaulting to plain new ex aws config and calling a function with arity 2 from there? e.g.:

````elixir
def presigned(%ExAws.S3.DirectUpload{} = upload) do
  presigned(upload, ExAws.Config.new(:s3)
end
def presigned(%ExAws.S3.DirectUpload{} = upload, %{} = config) do
    %{
      url: url(upload, config),
      credentials: credentials(upload, config)
    }
end
````
and after that some internal functions could be cleaned up a bit like:
````elixir
defp security_token(config) do
  Map.get(config, :security_token)
end

# or even shorter ... don't know that the style preference is for you
defp region(config), do: Map.get(config, :region)
````

@psteininger and @halostatue  it would be great if we could move this forward!

halostatue commented:
It’s not less multitenancy, it’s that we generally use machine keys in production. This affects `dev` environments _only_ for our developers.

psteininger commented:
@halostatue I just noticed this PR. I haven't used this code in over a year, and notifications went to nowhere. I will be getting my hands dirty again with this in the next week. My apologies for crazy delay

Commit:
Allow use of ex_aws_s3 ~> 2.2

Testing Notes
=============

Tested with Erlang/OTP 23 and Elixir 1.11 successfully.

Trying with Erlang/OTP 24 and Elixir 1.12 failed due to changes in the
OTP crypto library.

** (UndefinedFunctionError) function :crypto.hmac/3 is undefined or private, use crypto:mac/4 instead

Technical change notes
=====================

This allows software which depends on this to update libraries.

I didn't notice anything alarming in https://diff.hex.pm/diff/ex_aws_s3/2.0.1..2.2.0

halostatue committed:
Merge pull request #1 from mikestok/optional_config

Allow use of ex_aws_s3 ~> 2.2

Commit:
Allow the use of Erlang/OTP 24

Technical change notes
=====================

This whichever of :crypto.mac/4 or :crypto.hmac/3 is available so that
we can roll forward to Erlang/OTP 24 (which removes :crypto.hmac/3)

halostatue committed:
Merge pull request #2 from mikestok/optional_config

Allow the use of Erlang/OTP 24

Commit:
Allow this to be used in projects with poison 5.0

Technical change notes
======================

Manual testing of our platform app didn’t reveal any errors.

halostatue committed:
Merge pull request #3 from mikestok/allow-poison-5

Allow this to be used in projects with poison 5.0

halostatue committed:
Pre-archive commit
